src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -564,6 +566,47 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
         }
     }
 
+    fn discover_crate_type(&self, crate_type: &str) -> CargoResult<Option<(String, String)>> {
+        let mut process = self.config.rustc()?.process();
+        process.arg("-")
+               .arg("--crate-name").arg("___")
+               .arg("--print=file-names")
+               .env_remove("RUST_LOG");
+
+        process.arg("--crate-type").arg(crate_type);
+
+        let output = process.exec_with_output().chain_err(|| {
+            format!("failed to run `rustc` to learn about \
+                     crate-type {} information", crate_type)
+        })?;
+
+        let error = str::from_utf8(&output.stderr).unwrap();
+        let output = str::from_utf8(&output.stdout).unwrap();
+        let mut lines = output.lines();
+        let not_supported = error.lines().any(|line| {
+            (line.contains("unsupported crate type") ||
+             line.contains("unknown crate type")) &&
+            line.contains(crate_type)
+        });
+        if not_supported {
+            return Ok(None);
+        }
+        let line = match lines.next() {
+            Some(line) => line,
+            None => bail!("malformed output when learning about \
+                crate-type {} information", crate_type),
+        };
+        let mut parts = line.trim().split("___");
+        let prefix = parts.next().unwrap();
+        let suffix = match parts.next() {
+            Some(part) => part,
+            None => bail!("output of --print=file-names has changed in \
+                the compiler, cannot parse"),
+        };
+
+        Ok(Some((prefix.to_string(), suffix.to_string())))
```

Comment:
Most of this function is copied from `probe_target_info_kind` but I don't think we can do much about that -- I don't see an easy way to deduplicate much of it in a clean way. Open to suggestions though.